Apex callback refactoring

khuck opened the pull request:
Fixes #4199

## Proposed Changes

  - libhpx no longer contains any APEX symbols.
  - Any libraries or plugins or executables that *do* depend on APEX will still link with it
  - APEX now only has one dependency (or two, if HPXMP is used) - libhpx

## Any background context you want to provide?

This eliminates circular dependencies between APEX and HPX, and is the first step to making APEX a module in HPX.

msimberg commented on hpx/util/external_timer.hpp:
```diff
@@ -0,0 +1,318 @@
+#define HPX_UTIL_WEAK_PRE __attribute__((weak))
+#define HPX_UTIL_WEAK_POST
+#endif
+#endif
```
Where are these needed/used? I don't see any signs of them here or in APEX.

khuck replied:
those are leftover from an aborted solution that wasn't needed (solved with HPX_EXPORT). I will remove them.

msimberg commented:
Thanks @khuck! Let me try to summarize your changes to see if I've understood them correctly:

- HPX now exposes an `external_timer` class which allows registering callbacks for various events. HPX doesn't call anything from APEX directly.
- APEX still depends on HPX the usual way. However, it is built as a shared library which initializes the callbacks through a global object in the APEX shared library. Dependent projects need to link to the APEX library, but this is handled through `target_link_libraries(hpx INTERFACE apex)` so that it's transparent to users.

khuck commented on examples/1d_stencil/CMakeLists.txt:
```diff
@@ -66,12 +66,13 @@ foreach(example_program ${example_programs})
 
 endforeach()
 
+
```
probably not

khuck commented on hpx/runtime/threads/thread_data_fwd.hpp:
```diff
@@ -24,11 +24,10 @@
 #if defined(HPX_HAVE_APEX)
-// forward declare the APEX object
-namespace apex {
+namespace hpx::util::external_timer {
```
will do

khuck commented on hpx/runtime/threads/thread_data_fwd.hpp:
```diff
@@ -24,11 +24,10 @@
     struct task_wrapper;
 }
-using apex_task_wrapper = std::shared_ptr<apex::task_wrapper>;
+using task_wrapper = std::shared_ptr<hpx::util::external_timer::task_wrapper>;
```
This should be removed - someone else added that "using" statement and I just updated what it defined.  I can remove it and be explicit instead.

khuck commented on hpx/util/external_timer.hpp:
```diff
@@ -0,0 +1,318 @@
+     * that wants to use this callback API needs to extend this class.
+     */
+    struct task_wrapper {
+        bool dummy;
```
no, I just assumed my struct needed some contents.  I'll try without.

khuck commented on src/CMakeLists.txt:
```diff
@@ -278,7 +278,12 @@ if (TARGET hpx::amplifier)
+  # we use the "INTERFACE" option for target_link_libraries.
+  # Because libhpx_apex is a shared object library, we don't need to specify
+  # the whole archive.
+  target_link_libraries(hpx INTERFACE hpx::apex)
```
I am not sure I understand.  You mean the `target_include_directories` on line 291?

msimberg replied:
Sorry, yes, I meant `target_include_directories`...

khuck commented on src/util/external_timer.cpp:
```diff
@@ -0,0 +1,131 @@
+    HPX_EXPORT update_task_address_t *update_task_address_function{nullptr};
+    HPX_EXPORT start_t *start_function{nullptr};
+    HPX_EXPORT stop_t *stop_function{nullptr};
+    HPX_EXPORT yield_t *yield_function{nullptr};
```
done

khuck commented:
> Thanks @khuck! Let me try to summarize your changes to see if I've understood them correctly:
> 
> * HPX now exposes an `external_timer` class which allows registering callbacks for various events. HPX doesn't call anything from APEX directly.
> * APEX still depends on HPX the usual way. However, it is built as a shared library which initializes the callbacks through a global object in the APEX shared library. Dependent projects need to link to the APEX library, but this is handled through `target_link_libraries(hpx INTERFACE apex)` so that it's transparent to users.

Yes, that sums it up nicely.  I'm addressing your review comments and will update the branch in a few...

khuck commented:
@msimberg ok, I addressed your comments and fixed two inspect errors that slipped through.  let me know if I need to address anything else...  the failing tests on (not required) APEX build are lated to apex policies not exiting at shutdown, those have been failing for a while now and aren't related to this PR.  I will have to fix them eventually, of course.

khuck commented:
I have no idea what Codacy is complaining about - those union members are used in the cpp file

hkaiser commented:
@khuck I marked the Codacy 'errors' as resolved an re-triggered the analysis. All should be fine now.

khuck commented:
Excellent, thanks!
